wcfsetup/install/files/lib/acp/form/LanguageAddForm.class.php

```diff
@@ -103,149 +75,123 @@ public function readParameters()
-
-        // copy content
-        LanguageEditor::copyLanguageContent($this->sourceLanguage->languageID, $this->language->languageID);
-
-        // reset caches
-        LanguageFactory::getInstance()->clearCache();
-        LanguageFactory::getInstance()->deleteLanguageCache();
-
-        $this->saved();
-
-        // show success message
-        WCF::getTPL()->assign([
-            'success' => true,
-            'objectEditLink' => LinkHandler::getInstance()->getControllerLink(
-                LanguageEditForm::class,
-                ['id' => $this->language->languageID]
-            ),
+        parent::createForm();
+
+        $locales = [
+            '' => WCF::getLanguage()->get('wcf.global.noSelection'),
+            ...$this->locales,
+        ];
+
+        $this->form->appendChildren([
+            FormContainer::create('data')
+                ->appendChildren([
+                    TextFormField::create('languageName')
+                        ->label('wcf.global.name')
+                        ->description('wcf.acp.language.name.description')
+                        ->maximumLength(255)
+                        ->required(),
+                    TextFormField::create('languageCode')
+                        ->label('wcf.acp.language.code')
+                        ->description('wcf.acp.language.code.description')
+                        ->maximumLength(20)
+                        ->required()
+                        ->addValidator(new FormFieldValidator('unique', function (TextFormField $formField) {
+                            if ($formField->getValidationErrors() !== []) {
+                                return;
+                            }
+
+                            if (
+                                $this->formObject instanceof Language
+                                && \mb_strtolower($this->formObject->languageCode) === $formField->getValue()
+                            ) {
+                                return;
+                            }
+
+                            if (LanguageFactory::getInstance()->getLanguageByCode($formField->getValue())) {
+                                $formField->addValidationError(new FormFieldValidationError(
+                                    'notUnique',
+                                    'wcf.acp.language.add.languageCode.error.notUnique'
+                                ));
+                            }
+                        }))
+                        ->addValidator(new FormFieldValidator(
+                            'locale',
+                            function (SingleSelectionFormField $formField) {
+                                if ($formField->getValidationErrors() !== []) {
+                                    return;
+                                }
+
+                                $localeField = $formField->getDocument()->getNodeById('locale');
+                                \assert($localeField instanceof SingleSelectionFormField);
```

Comment:
And here it should be `languageCode` field then. I suspect you initially tested with the validator attached to the `languageCode` field and then incompletely moved it over, because it belong to the `locale` field?

This full implementation appears to work for me (but please double check):

```
                        ->addValidator(new FormFieldValidator(
                            'locale',
                            static function (SingleSelectionFormField $formField) {
                                if ($formField->getValidationErrors() !== []) {
                                    return;
                                }

                                if ($formField->getValue() === '') {
                                    return;
                                }

                                $languageCodeField = $formField->getDocument()->getNodeById('languageCode');
                                \assert($languageCodeField instanceof TextFormField);

                                $test = \explode('_', $formField->getValue());
                                if (\count($test) === 1 || $languageCodeField->getValue() !== $test[0]) {
                                    $formField->addValidationError(
                                        new FormFieldValidationError(
                                            'localeMismatch',
                                            'wcf.acp.language.add.languageCode.error.localeMismatch',
                                            [
                                                'locale' => $formField->getValue(),
                                            ]
                                        )
                                    );
                                }
                            }
                        )),
```

Reply:
Oh, the error phrase name should contain `locale` instead of `languageCode`. That will need to be fixed in both the phrases and PHP.